### PR TITLE
Scripts: Add TypeScript support to linting command

### PR DESCRIPTION
**Pull request by sirreal**
## Description

With some TypeScript in Gutenberg, .ts and .tsx (TypeScript) files should undergo the similar formatting and linting to `.js` files.

## How has this been tested?

Change a .ts file, stage, and commit it.
Work on `@wordpress/react-i18n` and see that lint issues are reported correctly.

## Types of changes

Breaking change: Add lint setup and rules for TypeScript files.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: package.json
Comment (sirreal) [inline package.json:292]: This will be more involved, the [linter needs a TypeScript parser](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/README.md#configuration).

File: package-lock.json
Comment (gziolo) [inline package-lock.json:34476]: It isn't hoisted anymore, updating the lock file is the worst experience ever.

File: packages/eslint-plugin/package.json
Comment (sirreal) [inline packages/eslint-plugin/package.json:41]: The lockfile was updating in a way that broke when doing a simple `npm install` with this update dependencies. `eslint-plugin-jsdoc` was moving deeper and apparently couldn't be found when necessary (from scripts package?).

I manually changed the lockfile to keep the updated dependency in its current location, it seems to be stable.

File: packages/eslint-plugin/package.json
Comment (sirreal) [inline packages/eslint-plugin/package.json:35]: Parser is required for TS syntax.

File: packages/eslint-plugin/package.json
Comment (sirreal) [inline packages/eslint-plugin/package.json:34]: eslint does a poor job of linting TypeScript. For example:

```ts
export interface WPDataAttachedStore {
	getSelectors: () => WPDataFunctionArray;
	getActions: () => WPDataFunctionArray;
	subscribe: ( listener: () => void ) => () => void;
}
```

With `eslint:recommended`, the above errors with `no-unused-vars` because `listener` is named but not used. However, this is a correctly written interface and `listener` cannot be removed without losing information.

This prompted me to add `plugin:@typescript-eslint/eslint-recommended` [according to documentation](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin#recommended-configs).

File: packages/data/src/types.ts
Comment (sirreal) [inline packages/data/src/types.ts:2]: We'll see the linter will report errors here as part of the typescript-eslint recommended rules. [`Function` type is discouraged](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/ban-types.md#default-options):

> Avoid the `Function` type, as it provides little safety for the following reasons:
> - It provides no type safety when calling the value, which means it's easy to provide the wrong arguments.
> - It accepts class declarations, which will fail when called, as they are called without the new keyword.

Example:

```ts
function callAndToLower(f: Function) {
  f().toLowerCase(); // Unsound, may error: `f()` is of type `any`!
}
function safeCallAndToLower(f: () => string) {
  f().toLowerCase(); // All good!
}
// No type errors!
callAndToLower(class { }); // TypeError: Class constructors cannot be invoked without 'new' 
callAndToLower(() => null); // Runtime TypeError: Cannot read property 'toLowerCase' of null

// Sound type errors 👍
safeCallAndToLower(class { });
safeCallAndToLower(() => null);
safeCallAndToLower(() => "Hello!"); // "hello!"
```

FYI @youknowriad

File: package.json
Comment (sirreal) [inline package.json:290]: I think we could add `.tsx` as well. TS that includes JSX _requires_ the `.tsx` extension. I'd expect this to work seamlessly when and if a `.tsx` file is added and it comes at little or no cost now:
```suggestion
		"*.{js,ts,tsx}": [
```

File: package.json
Comment (sirreal) [inline package.json:297]:
```suggestion
		"packages/**/*.{js,ts,tsx}": [
```

File: package.json
Comment (sirreal) [reply package.json:290]: It looks like we need to ignore `.d.ts` - we're getting bad `eslint` errors for this `Unexpected var, use let or const instead. eslint(no-var)`:

https://github.com/WordPress/gutenberg/blob/3ec7aa5356dc947636d0e7426e3453ad5c654eab/typings/gutenberg-env/index.d.ts#L7

Comment (gziolo) [reply package.json:290]: What would be the case when defining types to use `tsx`? I'm mostly curious because this decision doesn't prevent someone from creating such a file so in theory, it's better to have it rather than not 😄

Comment (sirreal) [reply package.json:290]: Added `*.d.ts` to `.eslintignore`.

Comment (sirreal) [reply package.json:290]: > What would be the case when defining types to use tsx?

There's no reason to use `.tsx` extension unless the file will include JSX syntax. When strictly describing types, I don't think there's any reason to use `.tsx`. However, there's nothing in the system that prevents such a file from being added. It seems preferable to cover the possibility (maybe redundantly) instead of leaving it open.

File: packages/eslint-plugin/package.json
Comment (gziolo) [reply packages/eslint-plugin/package.json]: It should be all good after 3096429906baf3a001669c59e48527d9ea7a5fe2.

File: package.json
Comment (gziolo) [reply package.json:290]: I'm fine with your making the decision 👍

File: .eslintignore
Comment (gziolo) [inline .eslintignore:12]: It's fine if you are adding `*.d.ts` in Gutenberg, but it won't for other projects that use `@wordpress/eslint-plugin` or `@wordpress/scripts`. Now the question is how we can remove `*.d.ts` file from listing directly in ESLint plugin, any ideas?

Comment (sirreal) [reply .eslintignore:12]: Maybe they can be linted and this was overzealous. The issue I experienced seems to be this which has apparently been addressed: https://github.com/typescript-eslint/typescript-eslint/issues/2456 😕

**Comment by sirreal** [comment]
I fixed up some lint issues just to demo this. The when committing the react-i18n changes, lint-stages reports:

```
husky > pre-commit (node v14.15.5)
  ↓ Stashing changes... [skipped]
    → No partially staged files found...
  ❯ Running tasks...
    ↓ Running tasks for package-lock.json [skipped]
      → No staged files match package-lock.json
    ↓ Running tasks for packages/*/package.json [skipped]
      → No staged files match packages/*/package.json
    ↓ Running tasks for *.scss [skipped]
      → No staged files match *.scss
    ❯ Running tasks for *.{js,ts,tsx}
      ✔ wp-scripts format-js
      ✖ wp-scripts lint-js
    ↓ Running tasks for {docs/{toc.json,tool/*.js},packages/{*/README.md,components/src/*/**/README.md}} [skipped]
      → No staged files match {docs/{toc.json,tool/*.js},packages/{*/README.md,components/src/*/**/README.md}}
    ✔ Running tasks for packages/**/*.{js,ts,tsx}



✖ wp-scripts lint-js found some errors. Please fix them and try committing again.

…/gutenberg/packages/react-i18n/src/index.tsx
   30:1  error  Missing JSDoc @param "i18n" declaration     jsdoc/require-param
   48:1  error  Missing JSDoc @param "props" declaration    jsdoc/require-param
  116:0  error  Missing JSDoc @param "InnerComponent" type  jsdoc/require-param-type
  117:0  error  Missing JSDoc @return type                  jsdoc/require-returns-type

✖ 4 problems (4 errors, 0 warnings)
  2 errors and 0 warnings potentially fixable with the `--fix` option.
```

**Comment by sirreal** [comment]
This is failing because some existing TS files have lint errors. There seem to be a lot of formatting issues at least. I'd appreciate if anyone wants to help with the cleanup.

**Comment by gziolo** [comment]
> This is failing because some existing TS files have lint errors. There seem to be a lot of formatting issues at least. I'd appreciate if anyone wants to help with the cleanup.

How many errors are we talking?

Should we start with enabling linting for TS only in one of the packages to keep this PR small? Do you feel comfortable enabling TS linting in `@wordpress/scripts`. It's perfectly fine to add linting for files with TS extensions but the rest might wait. I'm just checking since it means breaking changes in two packages 😃

File: packages/eslint-plugin/configs/recommended.js
Comment (sirreal) [reply packages/eslint-plugin/configs/recommended.js:28]: Very much.
```suggestion
			// Don't require redundant JSDoc types in TypeScript files.
```

File: packages/react-i18n/src/index.tsx
Comment (sirreal) [reply packages/react-i18n/src/index.tsx:6]: I agree up to a point.

We looked at doing that and it was a nightmare at the time and would've required a lot of maintenance effort for little or no benefit. There's no way to work nicely with types in the module import/export system in JSDoc. We would've had to manually re-export every single type (with their type arguments!) one by one via JSDoc.

Now that we're allowing some TypeScript, it may be more feasible to re-export all the types, but I haven't looked in many months. It still won't be as easy as something like `export type * from 'react'` as far as I know.

Comment (gziolo) [reply packages/react-i18n/src/index.tsx:6]: Unless there is some way to let ESLint know that it should give itself a break when seeing `import type` when processing `no-restricted-imports` 🤔 

Let's see: https://github.com/eslint/eslint/issues/13758.

**Comment by sarayourfriend** [comment]
@sirreal There are a lot of unrelated linting errors in this PR right?

**Comment by gziolo** [comment]
> @sirreal There are a lot of unrelated linting errors in this PR right?

There are 47 errors reported and as far as I can tell they are valid issues with TS files.

**Comment by sarayourfriend** [comment]
> There are 47 errors reported and as far as I can tell they are valid issues with TS files.

Oh gotcha, I was mixing up errors and warnings! @sirreal I'm happy to tackle fixing the errors in the TS files.

File: packages/react-i18n/src/index.tsx
Comment (gziolo) [reply packages/react-i18n/src/index.tsx:6]: Maybe we should list React as peer dependency in that case? There are more packages that would require the same handling.

**Comment by sarayourfriend** [comment]
Hmmm, it looks like there's something funky going on here that I'm not quite sure of.

Following the trail of errors from `control-label/types` we get an error about not being able to resolve `../text/types`. If we go to that module, which exists at that correct path, we see that it has an error about not being able to resolve `../truncate/types`. That module also exists at that path and itself has no errors. I'm not sure why eslint is unable to resolve these moduels 🤔 

Also when I ran `npm install` on this branch I ended up with loads of `package-lock.json` changes.

**Comment by sirreal** [comment]
> I am seeing the following error:
> 
> > `'props' is defined but never used. eslint(no-unused-vars)`
> 
> However, `props` there is just a type annotation. Is something wrong with `@typescript-eslint/parser`?

It's being parsed correctly, but the error doesn't make any sense. This is eslint (for JavaScript) showing a nonsense lint error for something perfectly reasonable in TypeScript.

We can disable this rule for .ts + .tsx files (I've done that for a few already). I wonder if we can do anything smarter with the recommended configs to prevent silly things like this. I believe we're set up correctly according to the [docs here](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin#recommended-configs) but it's worth playing around.

File: packages/components/src/ui/control-group/types.ts
Comment (sirreal) [inline packages/components/src/ui/control-group/types.ts:5]: In my experience, these type imports can confuse babel (which Gutenberg uses to "compile" TypeScript sources). Babel may include the module in the dependency graph, but we only want types.
```suggestion
import type { CSSProperties } from 'react';
```

Theres a compiler option we should set to make sure this is prevented. We should set [`"importsNotUsedAsValues": "error"`](https://www.typescriptlang.org/tsconfig#importsNotUsedAsValues) in the extended tsconfig to prevent this.

File: packages/eslint-plugin/configs/recommended.js
Comment (gziolo) [inline packages/eslint-plugin/configs/recommended.js:17]: https://github.com/benmosher/eslint-plugin-import#importcore-modules

Cold we set `react` as core module for TS files to avoid having to set it as a package dependency or is it necessary for TS?

File: packages/eslint-plugin/configs/recommended.js
Comment (gziolo) [inline packages/eslint-plugin/configs/recommended.js:23]: Should it be limited to TS files? Like in:

https://github.com/benmosher/eslint-plugin-import#importparsers

File: packages/react-i18n/src/index.tsx
Comment (gziolo) [reply packages/react-i18n/src/index.tsx:6]: I reviewed all the changes again.

It won't scale with so many types required from React or Reakit 🙁

Should we at least disable this rule for all `packages/**/types.ts` files? Ideally, the rule shouldn't error for type imports in the first place.

**Review by gziolo — COMMENTED** [review]
It looks like it's close to ready 👏 Awesome work 💯

File: packages/eslint-plugin/configs/recommended.js
Comment (sarayourfriend) [reply packages/eslint-plugin/configs/recommended.js:17]: It's necessary for TS https://stackoverflow.com/questions/55198502/using-eslint-with-typescript-unable-to-resolve-path-to-module

File: packages/eslint-plugin/configs/recommended.js
Comment (sarayourfriend) [reply packages/eslint-plugin/configs/recommended.js:23]: I think we can just use a single parser right? `@typescript-eslint/parser` will parse regular JS files just fine, but it's up to you if you prefer to scope it that's fine too.

Comment (gziolo) [reply packages/eslint-plugin/configs/recommended.js:23]: I don't know. I hoped you would know better based on the experience in Calypso 😃

File: packages/eslint-plugin/configs/recommended.js
Comment (gziolo) [reply packages/eslint-plugin/configs/recommended.js:17]: The resolver for file extension is necessary. That's one thing. 

I'm wondering if you need to explicitly declare an import for a package with only types as a dependency of the package that does it. In this example, you import only types from the `react` package. Does `react` need to be declared as a dependency of the package for TypeScript to work correctly when the package is consumed from npm.

Comment (sarayourfriend) [reply packages/eslint-plugin/configs/recommended.js:17]: Ohhh sorry, I misunderstood your original question.

I think yes, we can set `react` in that option and remove it from peerDependencies.

File: packages/eslint-plugin/configs/recommended.js
Comment (sarayourfriend) [reply packages/eslint-plugin/configs/recommended.js:23]: Unfortunately I do not. Maybe @jsnajdr or @sirreal or @scinos would have some idea.

We might as well do it if there's an option for it 🤷🏼‍♀️

Comment (sirreal) [reply packages/eslint-plugin/configs/recommended.js:23]: 🤷 I don't know, we can try it. I don't think we have anything to use by letting this parser handle everything. I don't know whether there are any downsides.

File: packages/components/tsconfig.json
Comment (sirreal) [inline packages/components/tsconfig.json:4]: Let's move this to the base config at `tsconfig.base.json`. We want all packages to use it.

File: packages/eslint-plugin/configs/recommended.js
Comment (sarayourfriend) [reply packages/eslint-plugin/configs/recommended.js]: Using `import/parsers` didn't seem to work anyway. Reverting that change back to using a single parser for everything.

File: packages/react-i18n/src/index.tsx
Comment (sarayourfriend) [reply packages/react-i18n/src/index.tsx:6]: I think we should just disable those rules for `packages/components`... G2 imports a _lot_ from reakit as well, outside of typescript files.

**Comment by sarayourfriend** [comment]
I'm not sure how to fix this parsing error. The syntax the parser is choking on is supported by TypeScript and seems to parse fine in tests in ASTExplorer: https://astexplorer.net/#/gist/a787487b51e47563a777aaefc9d5f2fb/7bba95fef1b3846d611a31423ed28261038f14e7

Any ideas @sirreal?

**Comment by sirreal** [comment]
It's erroring in `packages/dependency-extraction-webpack-plugin/lib/types.d.ts`? Eslint always reports errors for me in `.d.ts` files. It looks like it should use the correct parser and not fail to parse, but 🤷 .

I don't think we should eslint lint there, it's not really an appropriate tool for linting declaration files. I think we should add `*.d.ts` it to the global ignore and/or exclude `*.d.ts` extensions from the lint commands somehow.

**Comment by sarayourfriend** [comment]
This is ready for review.

File: typings/gutenberg-env/index.d.ts
Comment (gziolo) [inline typings/gutenberg-env/index.d.ts:1]: It's probably no longer necessary now that `*.d.ts` files are ignored.

**Review by gziolo — APPROVED** [review]
Why do you think that TS support for both `@wordpress/eslint-plugin` and `@wordpress/scripts` it's a breaking change rather than a new feature?

I might ask it a few weeks earlier. It isn't that important but I'm thinking about what could go wrong 😄 

I personally don't have enough expertise to tell that everything is perfect, but it is good enough for me to start using it with Gutenberg. I would defer the final call to @sirreal or someone with a better overview of how TS should be integrated in the project.

File: packages/react-i18n/src/index.tsx
Comment (gziolo) [reply packages/react-i18n/src/index.tsx:6]: We could as well remove those checks for `react` and `reakit` if that becomes an issue. Although, my initial comment still applies. Technically speaking both those libraries are implementation details and shouldn't leak outside of `@wordpress/element` (`react`) and `@wordpress/components` (`reakit`). If there is no way to re-export the types that are used in other packages, then we shouldn't fight with the linter and disable the rule as a primary solution. We better remove the rule that becomes hostile.

**Comment by sarayourfriend** [comment]
> Why do you think that TS support for both @wordpress/eslint-plugin and @wordpress/scripts it's a breaking change rather than a new feature?

I'm not sure it is a breaking change, but I guess if someone had been using `@wordpress/eslint-plugin` and had already integrated TypeScript into their project, suddenly they'll start getting new errors to fix if they hadn't been linting TypeScript files. I think I'd classify that as a breaking change rather than just a new feature, at least if new features should be 'non-disruptive' this isn't that.

**Comment by gziolo** [comment]
Yes, it might be an excellent reason to be cautious in case someone has integrated TS support themselves. That makes sense 👍🏻

In fact, we won't be able to publish a new version of both packages to npm until WordPress 5.7 is out which is on Match 9th, so we still might add more breaking changes to the mix 😄

**Comment by sirreal** [comment]
My reasoning for the breaking change is that:
- We're introducing new rules.
- We're linting files that wouldn't be linted before (`*.ts`).

Either could mean that the upgrade requires some manual work to get this ready and is worth mentioning so folks no what to expect. Anything that may require work on the consumer's end to upgrade is my understanding of a breaking change.

> I personally don't have enough expertise to tell that everything is perfect, but it is good enough for me to start using it with Gutenberg.

I certainly don't know whether this is perfect, but it's at a point where it can start getting exercise. If we're blocked on package releases can we `next` package releases? Maybe we can start getting feedback from other projects. @nerrad may have know some projects interested in using some of the TypeScript lint/prettier functionality.

**Comment by sirreal** [comment]
I can't approve this because I opened it, although @sarayourfriend has done a lot of the work 🙇‍♂️ . I think it's ready to be merged and see how it performs.

**Comment by gziolo** [comment]
> If we're blocked on package releases can we next package releases?

Sure thing, we can do it as soon as there is the next Gutenberg plugin RC ready 👍🏻

**Comment by nerrad** [comment]
> I certainly don't know whether this is perfect, but it's at a point where it can start getting exercise. If we're blocked on package releases can we next package releases? Maybe we can start getting feedback from other projects. @nerrad may have know some projects interested in using some of the TypeScript lint/prettier functionality.

We're implementing TS in WooCommerce Blocks feature plugin and definitely could benefit from this update (which looks good afaict).

**Pull request merged**
